### PR TITLE
feat(cloud): add shared-instance flag in limit superflag in alpha #8513

### PR DESCRIPTION
**Pull request by anurags92**
This PR adds a shared-instance flag to --limit superflag. When set to true (false by default), it will:

- Restrict access to any of the ACL operations like Login, add/remove/update user from non-galaxy namespaces.
- Prevent the leaking of environment variables for minio and aws.

(cherry picked from commit 5f3cece75da375077c45ab64228cb7053cf03d02) (cherry picked from commit eeb7bea1a7cb49636d38c14994cf3cf10bf857d2)

### CODE REVIEW

**Comment by anurags92** [comment]
In lieu of --- https://github.com/dgraph-io/dgraph/pull/8513

**Comment by skrdgraph** [comment]
it's failing on acl tests

**Comment by anurags92** [comment]
> it's failing on acl tests
That's fixed. But now failing on other tests. I'll work on it and make it green.

File: codec/codec.go [hunk 1]
```diff
@@ -405,7 +405,8 @@ func Decode(pack *pb.UidPack, seek uint64) []uint64 {
 
 // DecodeToBuffer is the same as Decode but it returns a z.Buffer which is
 // calloc'ed and can be SHOULD be freed up by calling buffer.Release().
-func DecodeToBuffer(buf *z.Buffer, pack *pb.UidPack) {
```
Comment (mangalaman93) [inline codec/codec.go:408]: I removed this change from my PR. We don't need it

File: codec/codec.go [hunk 2]
```diff
@@ -416,6 +417,7 @@ func DecodeToBuffer(buf *z.Buffer, pack *pb.UidPack) {
 			last = u
 		}
 	}
+	return buf
```
Comment (mangalaman93) [inline codec/codec.go:420]: same

File: codec/codec_test.go [hunk 3]
```diff
@@ -75,7 +75,6 @@ func TestBufferUidPack(t *testing.T) {
 	// Some edge case tests.
 	pack := Encode([]uint64{}, 128)
 	FreePack(pack)
-
```
Comment (mangalaman93) [inline codec/codec_test.go:78]: please remove this change too

File: dgraph/cmd/alpha/run.go [hunk 4]
```diff
@@ -208,6 +208,10 @@ they form a Raft group and provide synchronous replication.
 				"worker in a failed state. Use -1 to retry infinitely.").
 		Flag("txn-abort-after", "Abort any pending transactions older than this duration."+
 			" The liveness of a transaction is determined by its last mutation.").
+		Flag("shared-instance", "When set to true, it disables ACLs for non-galaxy users. "+
+			"It expects the access JWT to be contructed outside dgraph for those users as even "+
```
Comment (mangalaman93) [inline dgraph/cmd/alpha/run.go:212]: constructed*
We need to improve the language. I am not sure what this line means.
`t expects the access JWT to be contructed outside dgraph for those users as even login is denied to them`

File: dgraph/cmd/increment/increment.go [hunk 5]
```diff
@@ -200,6 +200,7 @@ func run(conf *viper.Viper) {
 		dg = dgTmp
 	}
 
+
```
Comment (mangalaman93) [inline dgraph/cmd/increment/increment.go:203]: remove this change too

File: systest/multi-tenancy/basic_test.go [hunk 6]
```diff
@@ -47,9 +47,10 @@ var timeout = 5 * time.Second
 // this file can me made common to the other acl tests as well. Needs some refactoring as well.
 func TestAclBasic(t *testing.T) {
 	prepare(t)
-	galaxyToken := testutil.Login(t,
+	galaxyToken, err := testutil.Login(t,
```
Comment (mangalaman93) [inline systest/multi-tenancy/basic_test.go:50]: this is so useful. I have seen tests failing because the token is nil sometimes. Should we check for token being nil?

File: systest/cloud/minio.env [hunk 7]
```diff
@@ -0,0 +1,2 @@
+MINIO_ACCESS_KEY=accesskey
```
Comment (mangalaman93) [inline systest/cloud/minio.env:1]: there is a file that is already there in the dgraph folder. Can we just use that?

File: systest/cloud/cloud_test.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2021 Dgraph Labs, Inc. and Contributors
```
Comment (mangalaman93) [inline systest/cloud/cloud_test.go:2]: 2023

File: systest/cloud/cloud_test.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+	"github.com/dgraph-io/dgo/v210/protos/api"
+	"github.com/dgraph-io/dgraph/graphql/e2e/common"
+	"github.com/dgraph-io/dgraph/testutil"
+	"github.com/dgraph-io/dgraph/x"
+
+	"github.com/stretchr/testify/require"
+)
+
+func prepare(t *testing.T) {
```
Comment (mangalaman93) [inline systest/cloud/cloud_test.go:36]: setup is probably a better word!

File: systest/cloud/cloud_test.go [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+	require.Zero(t, len(resp.Errors))
+
+	// Galaxy guardian should provide the crednetials as well.
+	resp = testutil.Export(t, galaxyToken, minioDest, "accesskey", "secretkey")
+	require.Zero(t, len(resp.Errors))
+
+}
+
+func TestMain(m *testing.M) {
+	fmt.Printf("Using adminEndpoint : %s for cloud test.\n", testutil.AdminUrl())
```
Comment (mangalaman93) [inline systest/cloud/cloud_test.go:183]: we can call the setup/prepare function here instead of in the each test

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

File: dgraph/cmd/alpha/run.go (hunk 4 above)
Comment (mangalaman93) [reply dgraph/cmd/alpha/run.go:212]: This is still not addressed yet

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

File: dgraph/cmd/alpha/run.go (hunk 4 above)
Comment (anurags92) [reply dgraph/cmd/alpha/run.go:212]: I have changed the language (added another line just after. Let me know if you have a suggestion?

Comment (mangalaman93) [reply dgraph/cmd/alpha/run.go:212]: I think the purpose of this flag is not so that the customer use external JWT libraries, "This can be done via online tools or jwt libraries". I am not sure how this helps.

**Comment by anurags92** [comment]
Closing this in favour of - https://github.com/dgraph-io/dgraph/pull/8625
